### PR TITLE
Strip cdata from the value when using WebsiteAgent

### PR DESCRIPTION
**Pull request by oroce**
There are a few RSS feeds which contains `<![CDATA[]]` to include HTML content in their rss. (some, like github uses escaped html).

With this change you can extract the content from cdata tags. You just need to set `strip_cdata: true` in extract options. 

Example WebsiteAgent:
```
{
  "name": "feed with cdata",
  "type": "rss",
  "url: "<url>",
  "extract": {
     "embed_code": {
         "css": "feed entry embed",
         "value": ".//text()"
         "strip_cdata": true
      }  
  }
}
```

With the following RSS:
```
<feed>
  <item>
     <embed>
         <![CDATA[<iframe src="https://youtube.com/watch?v=video1"></iframe>]]
     </embed>
     <embed>
         <![CDATA[<iframe src="https://youtube.com/watch?v=video2"></iframe>]]
     </embed>
  </item>
</feed
```

The result is: 
```
[
   '<iframe src="https://youtube.com/watch?v=video1"></iframe>',
   '<iframe src="https://youtube.com/watch?v=video2"></iframe>'
]
```

We've been using this changed `WebsiteAgent` for couple of months to extract `<iframe src="..">` from CDATA and forward it to tumblr and wordpress.

RSpec and updated documentation attached.

### CODE REVIEW

**Comment by cantino** [comment]
Thanks @oroce!

@knu / @dsander, do you think we should just always do this?

**Comment by oroce** [comment]
I'm open to any suggestions, I just couldn't a better solution. (Xpath can be very messy if it's about stripping cdata)

**Comment by dsander** [comment]
@cantino Sounds like a good idea to always do it, but I don't know if there are use case where one would need the CDATA tags.

File: spec/models/agents/website_agent_spec.rb [hunk 1]
```diff
@@ -529,6 +529,41 @@
...
+        end
+
+        it "works with XPath" do
+          expect {
+            @checker.check
+          }.to change { Event.count }.by(10)
+          event = Event.last
+          expect(event.payload['author']).to eq('bill98')
+          expect(event.payload['title']).to eq('<![CDATA[Help: Rainmeter Skins • Test if Today is Between 2 Dates]]>')
+          expect(event.payload['content']).to start_with('<![CDATA[Can I ')
```
Comment (cantino) [inline spec/models/agents/website_agent_spec.rb]: Don't you want to test that it removes these correctly?

Comment (cantino) [inline spec/models/agents/website_agent_spec.rb]: Oh, never mind, I see the different cases now.

File: app/models/agents/website_agent.rb [hunk 2]
```diff
@@ -439,7 +441,14 @@ def extract_xml(doc)
         case nodes
         when Nokogiri::XML::NodeSet
           result = nodes.map { |node|
-            case value = node.xpath(extraction_details['value'] || '.')
+            value = node.xpath(extraction_details['value'] || '.')
+            if extraction_details['strip_cdata']
```
Comment (cantino) [inline app/models/agents/website_agent.rb]: I'm pretty tempted to say that we should make this whole block into the following: 

```ruby
value = node.xpath(extraction_details['value'] || '.')
value = value.first.text if value.first && value.first.cdata?
value = value.to_i if value.is_a?(Float) && value.to_i == value
value.to_s
```

this cleans up the single-case `case` statement and also always does cdata unwrapping.

**Comment by cantino** [comment]
@knu (or maybe @virtadpt or @irfancharania), can you think of any case where someone would want to keep literal CDATA blocks in the output?

**Comment by irfancharania** [comment]
I can't think of a case where I'd want to hang on to the `<![CDATA` part for the website agent. This is much better than my pulling content out via regex.

**Comment by virtadpt** [comment]
@cantino I can't.  I don't know why anybody would use WebsiteAgent to parse an RSS or ATOM feed, because RSSAgent does that already, and does it with a lot less parsing hackery.

**Comment by cantino** [comment]
@oroce, mind making this just always be the behavior, as will the code example I left above?

**Comment by oroce** [comment]
@cantino yeah sure

**Comment by oroce** [comment]
@cantino i submitted my own version, see a796caa

the oneliner you wrote well, we have a little bit more complex case, it should look like this:
```
value = node.xpath(extraction_details['value'] || '.')
value = value.first.text if value.is_a?(Nokogiri::XML::NodeSet) && value.first.is_a?(Nokogiri::XML::Node) && value.first.cdata?
value = value.to_i if value.is_a?(Float) && value.to_i == value
value.to_s
```

We can go either one, I'm neutral on this, but imho the latter one (the oneliner) is really hard to read and I didn't find an easier way unfortunately, because:
* `value.first` throws error when `value` is `Integer` or `Float` (nomethoderror)
* `value.first` is false positive for `String` because it does not have `cdata?` method
* there's a test case where `value.first` returns a `Nokogiri::XML::NodeSet` which does not have `cdata?` method
* everytime we are dealing with cdata, `value` will be `NodeSet` and `value.first` has to be `Node`

If you have more test cases for CDATA I would love to add them.

File: app/models/agents/website_agent.rb [hunk 3]
```diff
@@ -46,6 +46,8 @@ class WebsiteAgent < Agent
 
       Beware that when parsing an XML document (i.e. `type` is `xml`) using `xpath` expressions all namespaces are stripped from the document unless a toplevel option `use_namespaces` is set to true.
 
+      If the extracted value contains `<![CDATA[content]]` you can get the `content` part if you set `strip_cdata` to true.
```
Comment (cantino) [inline app/models/agents/website_agent.rb]: You can remove this info now.

**Comment by cantino** [comment]
I'm fine keeping your version, I agree that it's more readable.

File: app/models/agents/website_agent.rb (hunk 3 above)
Comment (oroce) [inline app/models/agents/website_agent.rb]: removed, see 38b81bc

**Comment by cantino** [comment]
Thanks @oroce!

**Comment by knu** [comment]
Oh, I think I completely missed this PR.  You should probably use the XPath expression `string(.)` to extract what's in a CDATA section instead of `.//text()`.  I already fixed the documentation in WebsiteAgent.

**Comment by knu** [comment]
It seems to me that the special treatment on a node set whose first node is CDATA is inappropriate because it would truncate all the following nodes when you might want to extract all nodes contatenated as raw HTML/XML rather than just the content of the CDATA.

If XPath expressions like `string(.)` solves the original problem, can we consider reverting this change?

**Comment by cantino** [comment]
I don't think I understand the subtleties of this enough, but it sounds like this might not be a great solution to have in place.  @oroce?

**Comment by oroce** [comment]
totally makes sense what @knu says.

I used text() because I haven't that edgecase, when the cdata isn't the only node but there are siblings.

@knu, can you make that change and update the test with an example you mentioned?

**Comment by knu** [comment]
@oroce Thank you for your confirmation!  Actually, the specs were altered to use `string(.)` in #1744 by me not knowing this change, so they already no longer rely on it, that is, the block starting with `if value.is_a?(Nokogiri::XML::NodeSet)`.

I think we can safely revert this after you update your XPath expressions.  Could you confirm that expressions like `string(.)` work for you?

**Comment by oroce** [comment]
@knu since the [tests are still passing](https://github.com/cantino/huginn/blob/master/spec/models/agents/website_agent_spec.rb#L874-L907) which were written by [me and partly you](https://github.com/cantino/huginn/blame/master/spec/models/agents/website_agent_spec.rb#L874-L907). I see no reason why we couldnt move to `string(.)`.

**Comment by knu** [comment]
OK, done!

**Pull request merged**